java/src/main/java/ai/rapids/cudf/HostColumnVector.java

```diff
@@ -0,0 +1,1282 @@
+ *  limitations under the License.
+ *
+ */
+
+package ai.rapids.cudf;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+/**
+ * Similar to a ColumnVector, but the data is stored in host memory and accessible directly from
+ * the JVM. This class holds references to off heap memory and is reference counted to know when
+ * to release it.  Call close to decrement the reference count when you are done with the column,
+ * and call incRefCount to increment the reference count.
+ */
+public final class HostColumnVector implements AutoCloseable {
+  /**
+   * The size in bytes of an offset entry
+   */
+  static final int OFFSET_SIZE = DType.INT32.sizeInBytes;
+  private static final Logger log = LoggerFactory.getLogger(HostColumnVector.class);
+
+  static {
+    NativeDepsLoader.loadNativeDeps();
+  }
+
+  private final OffHeapState offHeap;
+  private final DType type;
+  private long rows;
+  private Optional<Long> nullCount = Optional.empty();
+  private int refCount;
+
+  /**
+   * Create a new column vector with data populated on the host.
+   */
+  HostColumnVector(DType type, long rows, Optional<Long> nullCount,
+                   HostMemoryBuffer hostDataBuffer, HostMemoryBuffer hostValidityBuffer) {
+    this(type, rows, nullCount, hostDataBuffer, hostValidityBuffer, null);
+  }
+
+  /**
+   * Create a new column vector with data populated on the host.
+   * @param type               the type of the vector
+   * @param rows               the number of rows in the vector.
+   * @param nullCount          the number of nulls in the vector.
+   * @param hostDataBuffer     The host side data for the vector. In the case of STRING
+   *                           this is the string data stored as bytes.
+   * @param hostValidityBuffer Arrow-like validity buffer 1 bit per row, with padding for
+   *                           64-bit alignment.
+   * @param offsetBuffer       only valid for STRING this is the offsets into
+   *                           the hostDataBuffer indicating the start and end of a string
+   *                           entry. It should be (rows + 1) ints.
+   */
+  HostColumnVector(DType type, long rows, Optional<Long> nullCount,
+                   HostMemoryBuffer hostDataBuffer, HostMemoryBuffer hostValidityBuffer,
+                   HostMemoryBuffer offsetBuffer) {
+    if (nullCount.isPresent() && nullCount.get() > 0 && hostValidityBuffer == null) {
+      throw new IllegalStateException("Buffer cannot have a nullCount without a validity buffer");
+    }
+    if (type != DType.STRING) {
+      assert offsetBuffer == null : "offsets are only supported for STRING";
+    }
+    offHeap = new OffHeapState(hostDataBuffer, hostValidityBuffer, offsetBuffer);
+    MemoryCleaner.register(this, offHeap);
+    this.rows = rows;
+    this.nullCount = nullCount;
+    this.type = type;
+    synchronized(this) {
```

Comment:
can `refCount=0` be implicit given `increfCountInternal(true)`? I think that would remove this `synchronized`.

Reply:
This is another instance of synchronize-on-this-in-constructor that I missed and should be removed.